source/simulator/assembly.cc

```diff
@@ -121,251 +126,119 @@ namespace aspect
       }
-    else
-      {
-        aspect::Assemblers::StokesBoundaryTraction<dim> *stokes_boundary_traction
-          = new aspect::Assemblers::StokesBoundaryTraction<dim>();
 
-        assemblers->stokes_system_on_boundary_face.push_back(
-          std_cxx11::unique_ptr<aspect::Assemblers::StokesBoundaryTraction<dim> > (stokes_boundary_traction));
+    if (parameters.use_discontinuous_temperature_discretization)
+      {
+        assemblers->advection_system_assembler_on_face_properties[0].need_face_material_model_data = true;
+        assemblers->advection_system_assembler_on_face_properties[0].need_face_finite_element_evaluation = true;
       }
 
-    // add the terms necessary to normalize the pressure
-    if (do_pressure_rhs_compatibility_modification)
+    if (parameters.use_discontinuous_composition_discretization)
       {
-        if (parameters.include_melt_transport)
-          {
-            aspect::Assemblers::MeltPressureRHSCompatibilityModification<dim> *melt_pressure_RHS_modification
-              = new aspect::Assemblers::MeltPressureRHSCompatibilityModification<dim>();
-
-            assemblers->stokes_system.push_back(
-              std_cxx11::unique_ptr<aspect::Assemblers::MeltPressureRHSCompatibilityModification<dim> > (melt_pressure_RHS_modification));
-          }
-        else
+        for (unsigned int i = 1; i<=introspection.n_compositional_fields; ++i)
           {
-            aspect::Assemblers::StokesPressureRHSCompatibilityModification<dim> *stokes_pressure_RHS_modification
-              = new aspect::Assemblers::StokesPressureRHSCompatibilityModification<dim>();
-
-            assemblers->stokes_system.push_back(
-              std_cxx11::unique_ptr<aspect::Assemblers::StokesPressureRHSCompatibilityModification<dim> > (stokes_pressure_RHS_modification));
+            assemblers->advection_system_assembler_on_face_properties[i].need_face_material_model_data = true;
+            assemblers->advection_system_assembler_on_face_properties[i].need_face_finite_element_evaluation = true;
           }
       }
+  }
 
-    if (parameters.include_melt_transport)
-      {
-        aspect::Assemblers::MeltAdvectionSystem<dim> *melt_advection_system
-          = new aspect::Assemblers::MeltAdvectionSystem<dim>();
+  template <int dim>
+  void
+  Simulator<dim>::
+  set_assemblers ()
+  {
+    // first let the manager delete all existing assemblers:
+    assemblers->reset();
 
-        assemblers->advection_system.push_back(
-          std_cxx11::unique_ptr<aspect::Assemblers::MeltAdvectionSystem<dim> > (melt_advection_system));
+    assemblers->advection_system_assembler_properties.resize(1+introspection.n_compositional_fields);
+    assemblers->advection_system_assembler_on_face_properties.resize(1+introspection.n_compositional_fields);
 
+    // Set up the set of assemblers depending on the mode of equations
+    if (parameters.include_melt_transport)
```

Comment:
where do you currently handle the error if somebody uses melt and newton? Maybe write this as ``if(melt && !newton) ... else if (newton) ... else if (!newton && !melt) ... else error``?